src/builder/create_attachment.rs

```diff
@@ -284,3 +272,37 @@ impl<'a> EditAttachments<'a> {
         self.new_and_existing_attachments.is_empty()
     }
 }
+
+impl<'a> Serialize for EditAttachments<'a> {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        #[derive(Serialize)]
+        struct NewAttachment<'a> {
+            id: u64,
+            filename: &'a Cow<'static, str>,
+            description: &'a Option<Cow<'a, str>>,
+        }
+
+        // Instead of an `AttachmentId`, the `id` field for new attachments corresponds to the
+        // index of the new attachment in the multipart payload. The attachment data will be
+        // labeled with `files[{id}]` in the multipart body. See `Multipart::build_form`.
+        let mut id = 0;
+        let mut seq = serializer.serialize_seq(Some(self.new_and_existing_attachments.len()))?;
+        for attachment in &self.new_and_existing_attachments {
+            match attachment {
+                NewOrExisting::New(new_attachment) => {
+                    let attachment = NewAttachment {
+                        id,
+                        filename: &new_attachment.filename,
+                        description: &new_attachment.description,
+                    };
+                    id += 1;
+                    seq.serialize_element(&attachment)?;
+                },
+                NewOrExisting::Existing(existing_attachment) => {
+                    seq.serialize_element(existing_attachment)?;
+                },
+            }
+        }
```

Comment:
Can this be rewritten to use `Serializer::collect_seq`?

Reply:
I don't think so, considering the types we're serializing in different between the two match arms.

Reply:
Ah yes, I don't know how I missed that, this would be possible if we had Either but probably not. I'll merge and look at it later if I get round to it.